File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/public/extensions/LLM/RunAnywhereLoRA.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Public API for LoRA adapter management.
 * Delegates to the generated LoRA proto-byte ABI in C++.
 *
 * LoRA (Low-Rank Adaptation) adapters allow fine-tuning behavior
 * of a loaded base model without replacing it.
 */

package com.runanywhere.sdk.public.extensions

import ai.runanywhere.proto.v1.DownloadProgress
import ai.runanywhere.proto.v1.ErrorCategory
import ai.runanywhere.proto.v1.ErrorCode
import ai.runanywhere.proto.v1.ExpectedModelFiles
import ai.runanywhere.proto.v1.InferenceFramework
import ai.runanywhere.proto.v1.LoRAApplyResult
import ai.runanywhere.proto.v1.LoraAdapterCatalogEntry
import ai.runanywhere.proto.v1.LoraAdapterCatalogGetRequest
import ai.runanywhere.proto.v1.LoraAdapterCatalogGetResult
import ai.runanywhere.proto.v1.LoraAdapterCatalogListRequest
import ai.runanywhere.proto.v1.LoraAdapterCatalogListResult
import ai.runanywhere.proto.v1.LoraAdapterCatalogQuery
import ai.runanywhere.proto.v1.LoraAdapterDownloadCompletedRequest
import ai.runanywhere.proto.v1.LoraAdapterDownloadCompletedResult
import ai.runanywhere.proto.v1.LoraAdapterImportRequest
import ai.runanywhere.proto.v1.LoraAdapterImportResult
import ai.runanywhere.proto.v1.LoraCompatibilityResult
import ai.runanywhere.proto.v1.ModelArtifactType
import ai.runanywhere.proto.v1.ModelCategory
import ai.runanywhere.proto.v1.ModelFileDescriptor
import ai.runanywhere.proto.v1.ModelFileRole
import ai.runanywhere.proto.v1.ModelFormat
import ai.runanywhere.proto.v1.ModelGetRequest
import ai.runanywhere.proto.v1.ModelInfoMetadata
import ai.runanywhere.proto.v1.ModelSource
import ai.runanywhere.proto.v1.SingleFileArtifact
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeLoraRegistry
import com.runanywhere.sdk.foundation.errors.SDKException
import com.runanywhere.sdk.public.RunAnywhere
import com.runanywhere.sdk.public.types.RALoRAAdapterConfig
import com.runanywhere.sdk.public.types.RALoRAApplyRequest
import com.runanywhere.sdk.public.types.RALoRARemoveRequest
import com.runanywhere.sdk.public.types.RALoRAState
import com.runanywhere.sdk.public.types.RAModelInfo
import com.runanywhere.sdk.utils.getCurrentTimeMillis
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Stateless capability namespace for LoRA adapter management.
 *
 * Mirrors Swift's `RunAnywhere.lora` (`RunAnywhere+LoRA.swift`) surface 1:1.
 * Runtime operations (apply / remove / list / state / checkCompatibility) and
 * catalog operations (register / listCatalog / queryCatalog / getCatalogEntry /
 * markDownloadCompleted / markImportCompleted) all flow through generated
 * request / result types from `lora_options.proto`.
 */
interface LoRA {
    /** Apply one or more LoRA adapters to the currently loaded model. */
    suspend fun apply(request: RALoRAApplyRequest): LoRAApplyResult

    /**
     * Apply one registered catalog adapter to the currently loaded model.
     *
     * Preserves `entry.id` in the generated config so commons can validate
     * registered catalog adapters against the loaded base model.
     */
    suspend fun apply(
        entry: LoraAdapterCatalogEntry,
        localPath: String? = null,
        scale: Float? = null,
        replaceExisting: Boolean = false,
    ): LoRAApplyResult {
        val adapterPath = localPath ?: entry.local_path?.takeIf { it.isNotBlank() }
        if (adapterPath.isNullOrBlank()) {
            throw SDKException.invalidArgument("LoRA catalog adapter '${entry.id}' has no local path")
        }
        return apply(
            RALoRAApplyRequest(
                adapters =
                    listOf(
                        RALoRAAdapterConfig(
                            adapter_path = adapterPath,
                            adapter_id = entry.id.takeIf { it.isNotBlank() },
                            scale = scale ?: entry.default_scale.takeIf { it > 0f } ?: 1f,
                        ),
                    ),
                replace_existing = replaceExisting,
            ),
        )
    }

    /** Remove adapters by generated request semantics, including `clear_all`. */
    suspend fun remove(request: RALoRARemoveRequest): RALoRAState

    /** Get info about all currently loaded LoRA adapters. */
    suspend fun list(): RALoRAState

    /** Get the LoRA service state reported by commons. */
    suspend fun state(): RALoRAState

    /**
     * Check whether a LoRA adapter is compatible with the current base model.
     * Mirrors Swift: returns an incompatible result instead of throwing.
     */
    suspend fun checkCompatibility(config: RALoRAAdapterConfig): LoraCompatibilityResult

    /** Register a LoRA adapter from a full catalog entry. */
    suspend fun register(entry: LoraAdapterCatalogEntry): LoraAdapterCatalogEntry

    /**
     * Register both the generated LoRA catalog entry and its generated download
     * artifact record. This does not fetch bytes.
     */
    suspend fun registerArtifact(entry: LoraAdapterCatalogEntry): RAModelInfo

    /**
     * Download a LoRA adapter through the canonical model-download pipeline.
     *
     * One call registers the catalog entry + artifact, downloads with
     * resume/checksum/progress through commons, records completion in the LoRA
     * catalog, and returns the stable local adapter path.
     */
    suspend fun download(
        entry: LoraAdapterCatalogEntry,
        onProgress: (suspend (DownloadProgress) -> Unit)? = null,
    ): String

    /** List catalog entries using the generated catalog request/result ABI. */
    suspend fun listCatalog(
        request: LoraAdapterCatalogListRequest = LoraAdapterCatalogListRequest(),
    ): LoraAdapterCatalogListResult

    /** Query catalog entries using generated filter semantics owned by commons. */
    suspend fun queryCatalog(query: LoraAdapterCatalogQuery): LoraAdapterCatalogListResult

    /** Fetch one catalog entry by generated request semantics. */
    suspend fun getCatalogEntry(request: LoraAdapterCatalogGetRequest): LoraAdapterCatalogGetResult

    /** Persist native-reported completion state after the platform has fetched bytes. */
    suspend fun markDownloadCompleted(
        request: LoraAdapterDownloadCompletedRequest,
    ): LoraAdapterDownloadCompletedResult

    /**
     * Import a user-picked local adapter file into SDK-owned storage.
     *
     * Kotlin only resolves platform access (e.g. content URI to a readable
     * path) before calling; commons owns everything past the source path:
     * deterministic catalog matching, canonical placement, artifact registry
     * record + manifest persistence, and catalog completion for matched
     * entries. Apps apply the returned `local_path`; they never construct
     * on-disk paths themselves.
     */
    suspend fun importAdapter(sourcePath: String): LoraAdapterImportResult

    /**
     * Persist native-reported import completion state. Sets `imported = true`
     * and a default status message when not already populated, matching the
     * IDL contract for platform file-picker / import completion.
     */
    suspend fun markImportCompleted(
        request: LoraAdapterDownloadCompletedRequest,
    ): LoraAdapterDownloadCompletedResult {
        val patched =
            request.copy(
                imported = true,
                status_message =
                    request.status_message.ifBlank { "import completed" },
            )
        return markDownloadCompleted(patched)
    }

    /**
     * Get all LoRA adapters compatible with a specific model
     * (CANONICAL_API §3, mirrors Swift `adaptersForModel`).
     */
    suspend fun adaptersForModel(modelId: String): List<LoraAdapterCatalogEntry> {
        val result = queryCatalog(LoraAdapterCatalogQuery(model_id = modelId))
        if (!result.success) {
            throw SDKException.make(
                code = ErrorCode.ERROR_CODE_PROCESSING_FAILED,
                message = result.error_message.ifBlank { "LoRA catalog query failed" },
                category = ErrorCategory.ERROR_CATEGORY_INTERNAL,
                shouldLog = false,
            )
        }
        return result.entries
    }

    /**
     * Get all registered LoRA adapters
     * (CANONICAL_API §3, mirrors Swift `allRegistered`).
     */
    suspend fun allRegistered(): List<LoraAdapterCatalogEntry> {
        val result = listCatalog()
        if (!result.success) {
            throw SDKException.make(
                code = ErrorCode.ERROR_CODE_PROCESSING_FAILED,
                message = result.error_message.ifBlank { "LoRA catalog list failed" },
                category = ErrorCategory.ERROR_CATEGORY_INTERNAL,
                shouldLog = false,
            )
        }
        return result.entries
    }
}

private const val LORA_ARTIFACT_MODEL_ID_PREFIX = "lora-adapter:"
private const val LORA_ARTIFACT_TAG = "lora-adapter"

private suspend fun ensureLoraReady() {
    if (!RunAnywhere.isInitialized) {
        throw SDKException.notInitialized("SDK not initialized")
    }
    RunAnywhere.ensureServicesReady()
}

/**
 * Stable model-registry id used for a LoRA adapter artifact.
 *
 * The adapter remains a LoRA catalog entry for apply/remove semantics, while
 * its bytes are represented as a generated model artifact so download/storage
 * policy stays on the generated registry/download path.
 */
val LoraAdapterCatalogEntry.loraArtifactModelID: String
    get() =
        if (id.startsWith(LORA_ARTIFACT_MODEL_ID_PREFIX)) {
            id
        } else {
            "$LORA_ARTIFACT_MODEL_ID_PREFIX$id"
        }

/**
 * Convert a generated LoRA catalog entry into generated model-registry
 * metadata used by the generic generated download path. Catalog filtering and
 * completion state remain owned by the generated LoRA catalog ABI.
 */
fun LoraAdapterCatalogEntry.toLoraArtifactModelInfo(
    timestampUnixMs: Long = getCurrentTimeMillis(),
): RAModelInfo {
    val artifactFilename = filename.ifBlank { url.substringAfterLast('/').substringBefore('?') }
    val descriptor =
        ModelFileDescriptor(
            url = url,
            filename = artifactFilename,
            is_required = true,
            size_bytes = size_bytes.takeIf { it > 0 },
            role = ModelFileRole.MODEL_FILE_ROLE_COMPANION,
            checksum_sha256 = checksum_sha256,
        )
    val expectedFiles =
        ExpectedModelFiles(
            files = listOf(descriptor),
            required_patterns = listOf(artifactFilename),
            description = "LoRA adapter artifact",
        )
    val metadataTags =
        buildList {
            add(LORA_ARTIFACT_TAG)
            compatible_models.forEach { add("base-model:$it") }
            addAll(tags)
        }.distinct()

    return RAModelInfo(
        id = loraArtifactModelID,
        name = name,
        category = ModelCategory.MODEL_CATEGORY_UNSPECIFIED,
        format = ModelFormat.MODEL_FORMAT_GGUF,
        framework = InferenceFramework.INFERENCE_FRAMEWORK_UNSPECIFIED,
        download_url = url,
        download_size_bytes = size_bytes,
        supports_lora = false,
        source = ModelSource.MODEL_SOURCE_REMOTE,
        created_at_unix_ms = timestampUnixMs,
        updated_at_unix_ms = timestampUnixMs,
        checksum_sha256 = checksum_sha256,
        metadata =
            ModelInfoMetadata(
                description = description,
                author = author.orEmpty(),
                license = license.orEmpty(),
                tags = metadataTags,
            ),
        single_file =
            SingleFileArtifact(
                required_patterns = listOf(artifactFilename),
                expected_files = expectedFiles,
            ),
        artifact_type = ModelArtifactType.MODEL_ARTIFACT_TYPE_SINGLE_FILE,
        expected_files = expectedFiles,
        is_available = true,
    )
}

/**
 * JVM/Android backing object for [LoRA]. Stateless; all calls
 * delegate to [CppBridgeLoraRegistry] on `Dispatchers.IO`.
 */
internal object AndroidLoRA : LoRA {
    override suspend fun apply(request: RALoRAApplyRequest): LoRAApplyResult {
        ensureLoraReady()
        return withContext(Dispatchers.IO) {
            CppBridgeLoraRegistry.apply(request)
        }
    }

    override suspend fun remove(request: RALoRARemoveRequest): RALoRAState {
        ensureLoraReady()
        return withContext(Dispatchers.IO) {
            CppBridgeLoraRegistry.remove(request)
        }
    }

    override suspend fun list(): RALoRAState {
        ensureLoraReady()
        return withContext(Dispatchers.IO) {
            CppBridgeLoraRegistry.list(RALoRAState())
        }
    }

    override suspend fun state(): RALoRAState {
        ensureLoraReady()
        return withContext(Dispatchers.IO) {
            CppBridgeLoraRegistry.state(RALoRAState())
        }
    }

    override suspend fun checkCompatibility(config: RALoRAAdapterConfig): LoraCompatibilityResult =
        try {
            ensureLoraReady()
            withContext(Dispatchers.IO) {
                CppBridgeLoraRegistry.compatibility(config)
            }
        } catch (e: Exception) {
            LoraCompatibilityResult(
                is_compatible = false,
                error_message = e.message.orEmpty(),
            )
        }

    override suspend fun register(entry: LoraAdapterCatalogEntry): LoraAdapterCatalogEntry {
        ensureLoraReady()
        return withContext(Dispatchers.IO) {
            CppBridgeLoraRegistry.register(entry)
        }
    }

    override suspend fun registerArtifact(entry: LoraAdapterCatalogEntry): RAModelInfo {
        val registeredEntry = register(entry)
        val artifact = registeredEntry.toLoraArtifactModelInfo()
        registerModelInternal(artifact)
        return artifact
    }

    override suspend fun download(
        entry: LoraAdapterCatalogEntry,
        onProgress: (suspend (DownloadProgress) -> Unit)?,
    ): String {
        ensureLoraReady()
        val artifact = registerArtifact(entry)
        val finalProgress = RunAnywhere.downloadModel(artifact, onProgress = onProgress)

        var localPath = finalProgress.local_path
        if (localPath.isBlank()) {
            val lookup = RunAnywhere.getModel(ModelGetRequest(model_id = artifact.id))
            if (lookup.found) {
                localPath = lookup.model?.local_path.orEmpty()
            }
        }
        if (localPath.isBlank()) {
            throw SDKException.operation(
                "LoRA adapter '${entry.id}' downloaded but no local path was recorded",
            )
        }

        markDownloadCompleted(
            LoraAdapterDownloadCompletedRequest(
                adapter_id = entry.id,
                local_path = localPath,
            ),
        )
        return localPath
    }

    override suspend fun listCatalog(
        request: LoraAdapterCatalogListRequest,
    ): LoraAdapterCatalogListResult {
        ensureLoraReady()
        return withContext(Dispatchers.IO) {
            CppBridgeLoraRegistry.listCatalog(request)
        }
    }

    override suspend fun queryCatalog(query: LoraAdapterCatalogQuery): LoraAdapterCatalogListResult {
        ensureLoraReady()
        return withContext(Dispatchers.IO) {
            CppBridgeLoraRegistry.queryCatalog(query)
        }
    }

    override suspend fun getCatalogEntry(
        request: LoraAdapterCatalogGetRequest,
    ): LoraAdapterCatalogGetResult {
        ensureLoraReady()
        return withContext(Dispatchers.IO) {
            CppBridgeLoraRegistry.getCatalogEntry(request)
        }
    }

    override suspend fun markDownloadCompleted(
        request: LoraAdapterDownloadCompletedRequest,
    ): LoraAdapterDownloadCompletedResult {
        ensureLoraReady()
        return withContext(Dispatchers.IO) {
            CppBridgeLoraRegistry.markDownloadCompleted(request)
        }
    }

    override suspend fun importAdapter(sourcePath: String): LoraAdapterImportResult {
        ensureLoraReady()
        return withContext(Dispatchers.IO) {
            CppBridgeLoraRegistry.importAdapter(LoraAdapterImportRequest(source_path = sourcePath))
        }
    }
}

val RunAnywhere.lora: LoRA
    get() = AndroidLoRA

File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/public/capabilities/runanywhere_lora.dart
// SPDX-License-Identifier: Apache-2.0
//
// runanywhere_lora.dart — LoRA capability surface (canonical §3 namespace).
// Mirrors Swift `RunAnywhere.LoRA` and Kotlin `RunAnywhere.lora` (G-A7).
//
// Canonical runtime and catalog surface:
//   apply / remove / list / state / checkCompatibility /
//   register / listCatalog / queryCatalog / getCatalogEntry /
//   markDownloadCompleted / adaptersForModel / allRegistered

import 'package:runanywhere/foundation/errors/sdk_exception.dart';
import 'package:runanywhere/generated/download_service.pbenum.dart'
    show DownloadState;
import 'package:runanywhere/generated/errors.pbenum.dart'
    show ErrorCategory, ErrorCode;
import 'package:runanywhere/generated/lora_options.pb.dart';
import 'package:runanywhere/generated/model_types.pb.dart' as model_pb;
import 'package:runanywhere/generated/model_types.pbenum.dart'
    show
        InferenceFramework,
        ModelCategory,
        ModelFileRole,
        ModelFormat,
        ModelSource;
import 'package:runanywhere/native/dart_bridge.dart';
import 'package:runanywhere/native/dart_bridge_lora.dart';
import 'package:runanywhere/native/dart_bridge_model_registry.dart';
import 'package:runanywhere/public/capabilities/runanywhere_downloads.dart';

/// LoRA (Low-Rank Adaptation) capability surface.
///
/// Access via `RunAnywhere.lora`. Mirrors Swift
/// `RunAnywhere.LoRA` and Kotlin `RunAnywhere.lora`.
class RunAnywhereLoRACapability {
  RunAnywhereLoRACapability._();
  static final RunAnywhereLoRACapability _instance =
      RunAnywhereLoRACapability._();
  static RunAnywhereLoRACapability get shared => _instance;

  // --- Runtime adapter operations ----------------------------------------

  /// Apply one or more LoRA adapters to the current model.
  Future<LoRAApplyResult> apply(LoRAApplyRequest request) async {
    return DartBridgeLora.shared.apply(request);
  }

  /// Apply one registered catalog adapter to the current model.
  ///
  /// Preserves [entry.id] in the generated config so commons can validate
  /// registered catalog adapters against the loaded base model.
  Future<LoRAApplyResult> applyCatalogAdapter(
    LoraAdapterCatalogEntry entry, {
    String? localPath,
    double? scale,
    bool replaceExisting = false,
  }) async {
    final adapterPath =
        localPath ?? (entry.localPath.isNotEmpty ? entry.localPath : '');
    if (adapterPath.isEmpty) {
      throw SDKException.make(
        code: ErrorCode.ERROR_CODE_INVALID_ARGUMENT,
        message: "LoRA catalog adapter '${entry.id}' has no local path",
        category: ErrorCategory.ERROR_CATEGORY_INTERNAL,
      );
    }

    final effectiveScale =
        scale ??
        (entry.hasDefaultScale() && entry.defaultScale > 0
            ? entry.defaultScale
            : 1.0);
    return apply(
      LoRAApplyRequest(
        adapters: [
          LoRAAdapterConfig(
            adapterPath: adapterPath,
            adapterId: entry.id,
            scale: effectiveScale,
          ),
        ],
        replaceExisting: replaceExisting,
      ),
    );
  }

  /// Remove one or more LoRA adapters, or clear all adapters.
  Future<LoRAState> remove(LoRARemoveRequest request) async {
    return DartBridgeLora.shared.remove(request);
  }

  /// Currently loaded LoRA adapters.
  Future<LoRAState> list() async {
    return DartBridgeLora.shared.list();
  }

  /// LoRA service state reported by commons.
  Future<LoRAState> state() async {
    return DartBridgeLora.shared.state();
  }

  /// Whether the current backend supports the given adapter.
  Future<LoraCompatibilityResult> checkCompatibility(
    LoRAAdapterConfig config,
  ) async {
    return DartBridgeLora.shared.checkCompatibility(config);
  }

  // --- Catalog operations -----------------------------------------------

  /// Register a LoRA adapter in the global registry. Entry is
  /// deep-copied internally by C++.
  Future<LoraAdapterCatalogEntry> register(
    LoraAdapterCatalogEntry entry,
  ) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
    return DartBridgeLoraRegistry.shared.register(entry);
  }

  /// Generated-proto LoRA catalog list surface.
  Future<LoraAdapterCatalogListResult> listCatalog([
    LoraAdapterCatalogListRequest? request,
  ]) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
    return DartBridgeLoraRegistry.shared.listCatalog(
      request ?? LoraAdapterCatalogListRequest(),
    );
  }

  /// Generated-proto LoRA catalog query surface.
  Future<LoraAdapterCatalogListResult> queryCatalog(
    LoraAdapterCatalogQuery query,
  ) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
    return DartBridgeLoraRegistry.shared.queryCatalog(query);
  }

  /// Generated-proto LoRA catalog get surface.
  Future<LoraAdapterCatalogGetResult> getCatalogEntry(
    LoraAdapterCatalogGetRequest request,
  ) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
    return DartBridgeLoraRegistry.shared.getCatalogEntry(request);
  }

  /// Record native-owned download/import completion in the commons LoRA catalog.
  Future<LoraAdapterDownloadCompletedResult> markDownloadCompleted(
    LoraAdapterDownloadCompletedRequest request,
  ) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
    return DartBridgeLoraRegistry.shared.markDownloadCompleted(request);
  }

  /// Record native-reported LoRA adapter import completion in commons.
  ///
  /// Mirrors Swift `RunAnywhere.lora.markImportCompleted(_:)`. Uses the
  /// generated download-completed message with `imported = true`, matching the
  /// IDL contract for platform file-picker/import completion.
  Future<LoraAdapterDownloadCompletedResult> markImportCompleted(
    LoraAdapterDownloadCompletedRequest request,
  ) async {
    final importRequest = request.deepCopy()..imported = true;
    if (importRequest.statusMessage.isEmpty) {
      importRequest.statusMessage = 'import completed';
    }
    return markDownloadCompleted(importRequest);
  }

  /// Import a user-picked local adapter file into SDK-owned storage.
  ///
  /// Dart only resolves platform access (file-picker temp path) before
  /// calling; commons owns everything past the readable source path:
  /// deterministic catalog matching, canonical placement, artifact registry
  /// record + manifest persistence, and catalog completion for matched
  /// entries. Mirrors Swift `RunAnywhere.lora.importAdapter(from:)`.
  Future<LoraAdapterImportResult> importAdapter(String sourcePath) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
    return DartBridgeLoraRegistry.shared.importAdapter(
      LoraAdapterImportRequest(sourcePath: sourcePath),
    );
  }

  /// All registered LoRA adapters compatible with [modelId].
  Future<List<LoraAdapterCatalogEntry>> adaptersForModel(String modelId) async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
    return DartBridgeLoraRegistry.shared.getForModel(modelId);
  }

  /// All registered LoRA adapters.
  Future<List<LoraAdapterCatalogEntry>> allRegistered() async {
    if (!DartBridge.isInitialized) {
      throw SDKException.notInitialized();
    }
    return DartBridgeLoraRegistry.shared.getAll();
  }

  // --- SDK-owned artifact registration + download -------------------------
  // Mirrors Swift RunAnywhere+LoRADownload.swift:97-160.

  static const _loraArtifactModelIdPrefix = 'lora-adapter:';
  static const _loraArtifactTag = 'lora-adapter';

  /// Stable model-registry id used for an adapter's download artifact.
  String _loraArtifactModelId(LoraAdapterCatalogEntry entry) =>
      entry.id.startsWith(_loraArtifactModelIdPrefix)
      ? entry.id
      : '$_loraArtifactModelIdPrefix${entry.id}';

  /// Convert a catalog entry into model-registry metadata used by the
  /// generic download path. Catalog filtering and completion state remain
  /// owned by the LoRA catalog ABI. Mirrors Swift
  /// `RALoraAdapterCatalogEntry.toLoraArtifactModelInfo()`.
  model_pb.ModelInfo _toLoraArtifactModelInfo(LoraAdapterCatalogEntry entry) {
    final urlTail = entry.url.split('/').isNotEmpty
        ? entry.url.split('/').last
        : entry.url;
    final artifactFilename = entry.filename.isNotEmpty
        ? entry.filename
        : urlTail.split('?').first;

    final descriptor = model_pb.ModelFileDescriptor(
      role: ModelFileRole.MODEL_FILE_ROLE_COMPANION,
      url: entry.url,
      filename: artifactFilename,
      relativePath: artifactFilename,
      isRequired: true,
    );
    if (entry.sizeBytes > 0) descriptor.sizeBytes = entry.sizeBytes;
    if (entry.hasChecksumSha256() && entry.checksumSha256.isNotEmpty) {
      descriptor.checksumSha256 = entry.checksumSha256;
    }

    final expectedFiles = model_pb.ExpectedModelFiles(
      files: [descriptor],
      requiredPatterns: [artifactFilename],
      description: 'LoRA adapter artifact',
    );

    final seen = <String>{};
    final tags = <String>[
      _loraArtifactTag,
      ...entry.compatibleModels.map((m) => 'base-model:$m'),
      ...entry.tags,
    ].where(seen.add).toList(growable: false);

    final metadata = model_pb.ModelInfoMetadata(
      description: entry.description,
      tags: tags,
    );
    if (entry.hasAuthor()) metadata.author = entry.author;
    if (entry.hasLicense()) metadata.license = entry.license;

    final model = model_pb.ModelInfo(
      id: _loraArtifactModelId(entry),
      name: entry.name,
      category: ModelCategory.MODEL_CATEGORY_UNSPECIFIED,
      format: ModelFormat.MODEL_FORMAT_GGUF,
      framework: InferenceFramework.INFERENCE_FRAMEWORK_UNKNOWN,
      downloadUrl: entry.url,
      source: ModelSource.MODEL_SOURCE_REMOTE,
      singleFile: model_pb.SingleFileArtifact(
        requiredPatterns: [artifactFilename],
        expectedFiles: expectedFiles,
      ),
      expectedFiles: expectedFiles,
      metadata: metadata,
      isAvailable: true,
    );
    if (entry.sizeBytes > 0) model.downloadSizeBytes = entry.sizeBytes;
    if (entry.hasChecksumSha256() && entry.checksumSha256.isNotEmpty) {
      model.checksumSha256 = entry.checksumSha256;
    }
    return model;
  }

  /// Register both the LoRA catalog entry and its downloadable artifact
  /// record. Does not fetch bytes. Mirrors Swift `lora.registerArtifact(_:)`
  /// (RunAnywhere+LoRADownload.swift:97).
  Future<model_pb.ModelInfo> registerArtifact(
    LoraAdapterCatalogEntry entry,
  ) async {
    final registered = await register(entry);
    final artifact = _toLoraArtifactModelInfo(registered);
    final saved = await DartBridgeModelRegistry.instance.saveProtoModel(
      artifact,
    );
    if (!saved) {
      // Mirrors Swift CppBridge.ModelRegistry.save failure
      // (CppBridge+ModelRegistry.swift:160-162): `.processingFailed`.
      throw SDKException.processingFailed(
        'Failed to save model via proto registry: ${artifact.id}',
      );
    }
    return artifact;
  }

  /// Download a LoRA adapter through the canonical model-download pipeline.
  ///
  /// One call does everything: registers the catalog entry + artifact,
  /// downloads with resume/checksum/progress via commons, records completion
  /// in the LoRA catalog, and returns the stable local path of the adapter
  /// file. Mirrors Swift `lora.download(_:onProgress:)`
  /// (RunAnywhere+LoRADownload.swift:111).
  Future<String> download(
    LoraAdapterCatalogEntry entry, {
    void Function(double progress)? onProgress,
  }) async {
    final artifact = await registerArtifact(entry);

    String localPath = '';
    await for (final progress in RunAnywhereDownloads.shared.start(
      artifact.id,
    )) {
      onProgress?.call(progress.overallProgress);
      if (progress.state == DownloadState.DOWNLOAD_STATE_FAILED) {
        // Mirrors Swift's `RunAnywhere.downloadModel` terminal-failure throw
        // (`.downloadFailed`, network category).
        throw SDKException.make(
          code: ErrorCode.ERROR_CODE_DOWNLOAD_FAILED,
          message: progress.errorMessage.isNotEmpty
              ? progress.errorMessage
              : 'LoRA adapter download failed for ${entry.id}',
          category: ErrorCategory.ERROR_CATEGORY_NETWORK,
        );
      }
      if (progress.state == DownloadState.DOWNLOAD_STATE_COMPLETED) {
        localPath = progress.localPath;
        break;
      }
    }

    if (localPath.isEmpty) {
      // The import step persisted the path on the registry record.
      final lookup = await DartBridgeModelRegistry.instance.getProtoModel(
        artifact.id,
      );
      localPath = lookup?.localPath ?? '';
    }
    if (localPath.isEmpty) {
      // Mirrors Swift RunAnywhere+LoRADownload.swift:128-134:
      // `.downloadFailed`, network category.
      throw SDKException.make(
        code: ErrorCode.ERROR_CODE_DOWNLOAD_FAILED,
        message:
            "LoRA adapter '${entry.id}' downloaded but no local path was recorded",
        category: ErrorCategory.ERROR_CATEGORY_NETWORK,
      );
    }

    await markDownloadCompleted(
      LoraAdapterDownloadCompletedRequest(
        adapterId: entry.id,
        localPath: localPath,
      ),
    );
    return localPath;
  }
}
